README.md
# Knot Installer

This script installs and configures a full [Knot System](https://github.com/knot-system), consisting of the modules [Knot Site](https://github.com/knot-system/knot-site), [Knot Home](https://github.com/knot-system/knot-home), [Knot Daemon](https://github.com/knot-system/knot-daemon), [Knot Auth](https://github.com/knot-system/knot-auth) and [Knot Control](https://github.com/knot-system/knot-control)

**This is an early beta version!** Some things may break, or change in the future!

## Initial Setup

Your server needs to run at least PHP 8.0 or later.

Download the latest `knot-install.php` and upload it to a folder on your webserver. Point your webbrowser to the `knot-install.php` file (for example, `https://www.example.com/knot-install.php`) - make sure, that the URL you open is the URL you want to use as your home page; if you want to use `https://`, open the `knot-install.php` with `https://`, and if you want to use a `www`-subdomain, make sure to include it. The `knot-install.php` will then guide you through the setup process, which will install and configure all modules in their own subfolders. After the installation, the `knot-install.php` gets deleted automatically.

You then can open the website at `https://www.example.com`.

The Knot Home backend can be reached inside the `knot-home/` subfolder (for example, `https://www.example.com/knot-home/`); there you can log in with your root URL (`https://www.example.com`).

For configuring and updating the Knot System install, log into the Knot Control backend at the `knot-control/` subfolder (for example, `https://www.example.com/knot-control/`); there you can log in with your root URL (`https://www.example.com`).


## Backup

You can export the sites you follow in Knot Home, when managing feeds.

To backup your Knot Site posts, copy the folder `knot-site/content/` to a safe location.

To backup the configuration, copy the `config.php` files inside the `knot-site/`, `knot-home/`, `knot-daemon/`, `knot-auth/` and `knot-control` subfolders.

For detailed instructions, read the `README.md` inside the `knot-site/`, `knot-home/`, `knot-daemon/`, `knot-auth/` and `knot-control/` subfolders.


## Changing the domain name

When you want to change the domain name (or move to a new domain), you need to update the domain at these places:

- in the `knot-site/config.php` update the URL at the `baseurl_overwrite` option. Also update the `mircosub` option (but keep the `/knot-daemon/` at the end)
- in `knot-daemon/config.php` update the URL in the `allowed_urls` option

if you change the (or move to or from a) subfolder, you also need to update the subfolder in these places:

- in the `knot-site/config.php` update the `basefolder_overwrite` option
- update the `RewriteBase` option in the `.htaccess`, `knot-site/.htaccess`, `knot-home/.htaccess`, `knot-daemon/.htaccess`, `knot-auth/.htaccess` and `knot-control/.htaccess` files


changelog.txt
v.0.2.1

- bugfix: remove knot-install.php after installation
- moved repo to 'knot-system' organization on GitHub


v.0.2.0

- renamed to 'Knot Installer' (was: 'Homestead Installer')
- updated GitHub urls and names of the knot modules


v.0.1.6

- this module is now named 'homestead-installer' (was only 'homestead')
- on install, set indieauth-metadata for eigenheim, instead of authorization_endpoint and token_endpoint
- moved all relevant code for the installer into one php file
- also install new module homestead-control
- removed updater, this is now part of the new 'homestead-control' module


v.0.1.5

- install 'einwohnermeldeamt' as the authorization & token endpoint
- bugfixes & enhancements


v.0.1.4

- better compatibility for some server configurations
- bugfixes & enhancements

Installer
- check, if some required extensions (like GD, CURL, SimpleXML, DOM) are installed
- check, if the config file was created

Updater
- only update selected modules, instead of all modules


v.0.1.3

Installer
- allow installing of unstable dev version
- add comments to .htaccess, at beginning and end of our block
- bugfixes & enhancements


v.0.1.2

Updater
- fix for displaying relevant release information for Homestead update


v.0.1.1

Updater
- added basic updater to update Homestead or the installed modules (the modules need to be at least at version 0.1.1 for the version-select-dropdown to work)

Installer
- added instructions about changing the domain


v.0.1.0

- initial release


config.php
<?php

$sources = [
	'eigenheim' => [
		'name' => 'Eigenheim',
		'target' => 'eigenheim/',
		'zipball_stable' => 'https://api.github.com/repos/maxhaesslein/eigenheim/releases',
		'zipball_dev' => 'https://github.com/maxhaesslein/eigenheim/archive/refs/heads/main.zip',
	],
	'sekretaer' => [
		'name' => 'Sekretär',
		'target' => 'sekretaer/',
		'zipball_stable' => 'https://api.github.com/repos/maxhaesslein/sekretaer/releases',
		'zipball_dev' => 'https://github.com/maxhaesslein/sekretaer/archive/refs/heads/main.zip',
	],
	'postamt' => [
		'name' => 'Postamt',
		'target' => 'postamt/',
		'zipball_stable' => 'https://api.github.com/repos/maxhaesslein/postamt/releases',
		'zipball_dev' => 'https://github.com/maxhaesslein/postamt/archive/refs/heads/main.zip',
	],
	'einwohnermeldeamt' => [
		'name' => 'Einwohnermeldeamt',
		'target' => 'einwohnermeldeamt/',
		'zipball_stable' => 'https://api.github.com/repos/maxhaesslein/einwohnermeldeamt/releases',
		'zipball_dev' => 'https://github.com/maxhaesslein/einwohnermeldeamt/archive/refs/heads/main.zip',
	]
];


update.php
<?php

include_once( 'config.php' );


if( ! isset($_REQUEST['debug']) ) {
	error_reporting(0);
}


$api_url = 'https://api.github.com/repos/maxhaesslein/homestead/releases';
$dev_zip = 'https://github.com/maxhaesslein/homestead/archive/refs/heads/main.zip';

$homestead_version = file_get_contents('version.txt');

$useragent = 'maxhaesslein/homestead/'.$homestead_version;


$self = basename(__FILE__);

$abspath = realpath(dirname(__FILE__)).'/';
$abspath = preg_replace( '/system\/$/', '', $abspath );

$basefolder = str_replace( $self, '', $_SERVER['PHP_SELF']);

if( isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] === 'on' ) $baseurl = 'https://';
else $baseurl = 'http://';
$baseurl .= $_SERVER['HTTP_HOST'];
$baseurl .= $basefolder;


$homestead_abspath = $abspath;
$homestead_baseurl = $baseurl;


$module_found = false; // at least one module was found
foreach( $sources as $source => $options ) {

	$sources[$source]['installed_version'] = false;
	if( is_dir($abspath.$options['target'].'/') && file_exists($abspath.$options['target'].'/system/version.txt') ) {
		$sources[$source]['installed_version'] = trim(file_get_contents($abspath.$options['target'].'/system/version.txt'));

		$module_found = true;
	}

}


$update_allowed = false;
if( file_exists($abspath.'update') || file_exists($abspath.'update.txt') ) {
	$update_allowed = true;
}


?><!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8">
	<meta name="viewport" content="width=device-width,initial-scale=1.0">
	<title>🏡 Homestead Updater</title>

	<style>
		fieldset {
			margin-top: 1.5em;
		}
			fieldset legend {
				font-weight: bolder;
			}
		fieldset.unimportant {
			opacity: 0.5;
		}
	hr {
		margin: 2em 0;
		border: 0;
		border-top: 1px solid;
	}
	code.response {
		border: 1px solid grey;
		padding: 10px;
		display: block;
		pointer-events: none;
		opacity: 0.5;
	}
		code.response a {
			text-decoration: none;
			color: inherit;
			pointer-events: none;
		}
	</style>
</head>
<body>
<main style="max-width: 600px; margin: 0 auto">

	<h1>🏡 Homestead Updater</h1>

<?php

if( ! $module_found ) {
	
	echo '<p>It looks like Homestead is not installed at this location!</p>';

} elseif( ! $update_allowed ) {
	?>
	<p>Please create an empty file called <code>update</code> (or <code>update.txt</code>) in the root folder (so we are sure that you have access to the webserver), and then refresh this page</p>
	<p><a href="<?= $baseurl ?>update.php">refresh this page</a></p>
	<?php

} elseif( isset($_GET['action']) && $_GET['action'] == 'update_homestead' ) {

	if( (isset($_GET['step']) && $_GET['step'] == 'install') ) {

		if( empty($_REQUEST['version']) || $_REQUEST['version'] == 'latest' ) {

			$json = get_remote_json( $api_url );
			
			if( ! $json || ! is_array($json) ) {
				?>
				<p><strong>Error:</strong> could not get release information from GitHub</p>
				<?php
				exit;
			}

			$latest_release = $json[0];

			$zipball = $latest_release->zipball_url;

			$zip_folder_name_start = 'maxhaesslein-homestead-';

		} elseif( $_REQUEST['version'] == 'dev' ) {

			$zipball = $dev_zip;

			$zip_folder_name_start = 'homestead-';

		} else {

			?>
			<p><strong>Error:</strong> unknown version</p>
			<?php
			exit;

		}

		if( ! $zipball ) {
			?>
			<p><strong>Error:</strong> could not get new .zip file from GitHub</p>
			<?php
			exit;
		}

		$temp_folder = $abspath.'_homestead_temp/';
		if( ! is_dir($temp_folder) ) {
			$oldumask = umask(0); // we need this for permissions of mkdir to be set correctly
			if( mkdir( $temp_folder, 0777, true ) === false ) {
				echo '<p><strong>Error:</strong> could not create temp folder</p>';
				exit;
			}
			umask($oldumask); // we need this after changing permissions with mkdir
		}

		echo '<p>Downloading new .zip from GitHub … ';
		flush();

		$temp_zip_file = $temp_folder.'_new_release.zip';
		if( file_exists($temp_zip_file) ) unlink($temp_zip_file);

		$file_handle = fopen( $temp_zip_file, 'w+' );

		$ch = curl_init( $zipball );
		curl_setopt( $ch, CURLOPT_FOLLOWLOCATION, 1 );
		curl_setopt( $ch, CURLOPT_RETURNTRANSFER, 1 );
		curl_setopt( $ch, CURLOPT_USERAGENT, $useragent );
		curl_setopt( $ch, CURLOPT_FILE, $file_handle );
		curl_exec( $ch );
		curl_close( $ch );

		fclose($file_handle);

		echo 'done.</p>';

		echo '<p>Extracting .zip file … ';
		flush();
		
		$temp_zip_folder = $temp_folder.'/_new_release/';
		if( is_dir($temp_zip_folder) ) deleteDirectory($temp_zip_folder);
		$oldumask = umask(0); // we need this for permissions of mkdir to be set correctly
		if( mkdir( $temp_zip_folder, 0777, true ) === false ) {
			echo '<p><strong>Error:</strong> could not create temp folder for zip unpacking</p>';
			exit;
		}
		umask($oldumask); // we need this after changing permissions with mkdir

		$zip = new ZipArchive;
		$res = $zip->open($temp_zip_file);
		if( $res !== TRUE ) {
			echo '<p><strong>Error:</strong> could not extract .zip file</p>';
			exit;
		}
		$zip->extractTo( $temp_zip_folder );
		$zip->close();

		echo 'done.</p>';

		$subfolder = false;
		foreach( scandir( $temp_zip_folder ) as $obj ) {
			if( $obj == '.' || $obj == '..' ) continue;
			if( ! is_dir($temp_zip_folder.$obj) ) continue;
			if( ! str_starts_with($obj, $zip_folder_name_start) ) continue;
			// the zip file should have exactly one subfolder. this is what we want to get here
			$subfolder = $temp_zip_folder.$obj.'/';
		}

		if( ! $subfolder ) {
			echo '<p><strong>Error:</strong> something went wrong with the .zip file</p>';
			exit;
		}

		echo '<p>Deleting old files … ';
		flush();

		// TODO: the new version should have a list of files to delete and copy, so that we can add new files with an update

		@unlink( $abspath.'.gitignore' );
		unlink( $abspath.'changelog.txt' );
		unlink( $abspath.'config.php' );
		unlink( $abspath.'install.php' );
		unlink( $abspath.'README.md' );
		unlink( $abspath.'update.php' );

		echo 'done.</p>';

		echo '<p>Moving new files to new location … ';
		flush();

		rename( $subfolder.'changelog.txt', $abspath.'changelog.txt' );
		rename( $subfolder.'config.php', $abspath.'config.php' );
		rename( $subfolder.'install.php', $abspath.'install.php' );
		rename( $subfolder.'README.md', $abspath.'README.md' );
		rename( $subfolder.'update.php', $abspath.'update.php' );

		echo 'done.</p>';
		echo '<p>Cleaning up …';
		@unlink( $abspath.'update.txt' );
		@unlink( $abspath.'update' );

		deleteDirectory( $temp_folder );

		echo 'done.</p>';

		echo '<p>Please <a href="'.$baseurl.'update.php">refresh this page</a></p>';

	} else {

		$json = get_remote_json( $api_url );
		
		if( ! $json || ! is_array($json) ) {
			?>
			<p><strong>Error:</strong> could not get release information from GitHub</p>
			<?php
			exit;
		}

		$latest_release = $json[0];

		$release_name = $latest_release->name;

		?>
		<p>Latest release: <strong><?= $release_name ?></strong><br>
		Currently installed: <strong><?= $homestead_version ?></strong></p>
		<?php

		$release_notes = array();

		$version_number_old = explode('.', $homestead_version);
		$version_number_new = explode('.', $release_name);

		if( $version_number_new[0] > $version_number_old[0] 
		 || ($version_number_new[0] == $version_number_old[0] && $version_number_new[1] > $version_number_old[1] )
		 || ($version_number_new[0] == $version_number_old[0] && $version_number_new[1] == $version_number_old[1] && $version_number_new[2] > $version_number_old[2] )
		){

			foreach( $json as $release ) {
				$tag_name = str_replace('v.', '', $release->tag_name);
				$release_number = explode('.', $tag_name);

				$newer_version = false;
				if( $release_number[0] > $version_number_old[0] 
				 || ($release_number[0] == $version_number_old[0] && $release_number[1] > $version_number_old[1] )
				 || ($release_number[0] == $version_number_old[0] && $release_number[1] == $version_number_old[1] && $release_number[2] > $version_number_old[2] )
				){
				 	$newer_version = true;
				}

				if( ! $newer_version ) break;

				$release_notes[] = [
					'title' => $release->tag_name,
					'body' => $release->body
				];
			}
		
			echo '<p><strong>New version available!</strong> You should update your system.</p>';

			if( count($release_notes) ) {
				?>
				<h2>Release notes:</h2>
				<?php
				
				foreach( $release_notes as $release_note ) {
					echo '<h3>'.$release_note['title'].'</h3>';

					$body = htmlentities($release_note['body']);
					$body = nl2br($body);
				
					echo $body;
				}

			}

		} else {
			echo '<p>You are running the latest version.</p>';
		}

		?>
		<hr>

		<form action="<?= $baseurl ?>update.php" method="GET">
			<input type="hidden" name="action" value="update_homestead">
			<input type="hidden" name="step" value="install">
			<p><label>Version: <select name="version">
				<option value="latest" selected>latest stable release (v.<?= $release_name ?>)</option>
				<option value="dev">unstable dev release (not recommended)</option>
			</select></label></p>
			<p><button>update Homestead system</button></p>
		</form>

		<?php

	}

} elseif( isset($_GET['action']) && $_GET['action'] == 'update_modules' ) {

	$version = 'latest';
	if( ! empty($_REQUEST['version']) ) $version = $_REQUEST['version'];

	$modules = [];
	if( ! empty($_REQUEST['modules']) ) $modules = $_REQUEST['modules'];

	if( ! count($modules) ) {
		?>
		<p>please select at least one module to update</p>
		<p><a href="<?= $baseurl ?>update.php">back</a></p>
		<?php
		exit;
	}

	?>
	<p>Starting update …</p>
	<?php
	flush();

	foreach( $_REQUEST['modules'] as $source ) {
		do_update( $source, $version );
	}

	echo '<p>Cleaning up …</p>';
	flush();

	@unlink($abspath.'update.txt');
	@unlink($abspath.'update');

	echo '<p>All done.</p>';

} else {
	?>
	
	<form method="GET" action="<?= $baseurl ?>update.php">

		<fieldset>
			<legend>Update Homestead</legend>
			
			<input type="hidden" name="action" value="update_homestead">

			<p><button>search for Homestead update</button></p>
		</fieldset>

	</form>


	<form method="GET" action="<?= $baseurl ?>update.php">

		<fieldset>
			<legend>Update Modules</legend>

			<p>This will update the selected modules. These modules are currently installed:</p>

			<input type="hidden" name="action" value="update_modules">

			<ul style="list-style-type: none; padding: 0;">
				<?php
				foreach( $sources as $source => $options ) {
					if( ! $options['installed_version'] ) continue;
					?>
					<li>
						<label><input type="checkbox" name="modules[]" value="<?= $source ?>" checked> <?= $options['name'] ?> (v.<?= $options['installed_version'] ?>)</label>
					</li>
					<?php
				}
				?>
			</ul>

			<p><label>update to: <select name="version">
				<option value="latest" selected>latest stable release</option>
				<option value="dev">unstable dev release (not recommended)</option>
			</select></label></p>
			<p><button>update selected modules</button></p>
		</fieldset>

	</form>
	<?php
}
?>
</main>
</body>
</html><?php


function do_update( $source, $version = 'latest' ) {

	global $homestead_abspath, $homestead_baseurl, $sources;

	$options = $sources[$source];

	echo '<p>Updating '.$options['name'].' …</p>';
	flush();

	$abspath = $homestead_abspath.$options['target'];
	$baseurl = $homestead_baseurl.$options['target'];

	touch( $abspath.'update' );

	time_nanosleep(0,500000000); // sleep for 0.5 seconds

	$response = get_request( $baseurl.'?update=true&step=install&version='.$version );

	?>
	<code class="response">
		<?php echo strip_tags( $response, ['br','p','ul','ol','li'] ); ?>
	</code>
	<?php
	flush();

	time_nanosleep(0,500000000); // sleep for 0.5 seconds

	get_request( $baseurl ); // trigger re-creation of missing files

	flush();

}


function get_request( $url ) {

	global $useragent;

	$ch = curl_init( $url );
	curl_setopt( $ch, CURLOPT_RETURNTRANSFER, 1 );
	curl_setopt( $ch, CURLOPT_USERAGENT, $useragent );
	curl_setopt( $ch, CURLOPT_FOLLOWLOCATION, true );
	$response = curl_exec( $ch );
	curl_close( $ch );

	return $response;
}

function get_remote_json( $url, $headers = array() ) {

	global $useragent;

	$ch = curl_init( $url );
	curl_setopt( $ch, CURLOPT_RETURNTRANSFER, 1 );
	curl_setopt( $ch, CURLOPT_USERAGENT, $useragent );
	curl_setopt( $ch, CURLOPT_FOLLOWLOCATION, true );
	$response = curl_exec($ch);
	curl_close( $ch );

	$json = json_decode($response);

	return $json;
}

function deleteDirectory( $dirPath ) {

	if( ! is_dir($dirPath) ) return;

	$objects = scandir($dirPath);
	foreach ($objects as $object) {
		if( $object == "." || $object == "..") continue;

		if( is_dir($dirPath . DIRECTORY_SEPARATOR . $object) ){
			deleteDirectory($dirPath . DIRECTORY_SEPARATOR . $object);
		} else {
			unlink($dirPath . DIRECTORY_SEPARATOR . $object);
		}
	}
	rmdir($dirPath);
}


knot-install.php
<?php

$installer_version = '0.2.1';


$sources = [
	'knot-site' => [
		'name' => 'Knot Site',
		'target' => 'knot-site/',
		'zipball_stable' => 'https://api.github.com/repos/knot-system/knot-site/releases',
		'zipball_dev' => 'https://github.com/knot-system/knot-site/archive/refs/heads/main.zip',
	],
	'knot-home' => [
		'name' => 'Knot Home',
		'target' => 'knot-home/',
		'zipball_stable' => 'https://api.github.com/repos/knot-system/knot-home/releases',
		'zipball_dev' => 'https://github.com/knot-system/knot-home/archive/refs/heads/main.zip',
	],
	'knot-daemon' => [
		'name' => 'Knot Daemon',
		'target' => 'knot-daemon/',
		'zipball_stable' => 'https://api.github.com/repos/knot-system/knot-daemon/releases',
		'zipball_dev' => 'https://github.com/knot-system/knot-daemon/archive/refs/heads/main.zip',
	],
	'knot-auth' => [
		'name' => 'Knot Auth',
		'target' => 'knot-auth/',
		'zipball_stable' => 'https://api.github.com/repos/knot-system/knot-auth/releases',
		'zipball_dev' => 'https://github.com/knot-system/knot-auth/archive/refs/heads/main.zip',
	],
	'knot-control' => [
		'name' => 'Knot Control',
		'target' => 'knot-control/',
		'zipball_stable' => 'https://api.github.com/repos/knot-system/knot-control/releases',
		'zipball_dev' => 'https://github.com/knot-system/knot-control/archive/refs/heads/main.zip',
	]
];



if( ! isset($_REQUEST['debug']) ) {
	error_reporting(0);
}


$php_min_version_major = 8;

$useragent = 'knot/installer/'.$installer_version;


$local_phpversion = explode( '.', phpversion() );

$self = basename(__FILE__);
$abspath = realpath(dirname(__FILE__)).'/';
$basefolder = str_replace( $self, '', $_SERVER['PHP_SELF']);

if( isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] === 'on' ) $baseurl = 'https://';
else $baseurl = 'http://';
$baseurl .= $_SERVER['HTTP_HOST'];
$baseurl .= $basefolder;


$is_installed = false;
foreach( $sources as $source => $options ) {
	if( is_dir($abspath.$options['target']) ) {
		$is_installed = true;
	}
}


$required_extensions = array(
	'gd',
	'simplexml',
	'curl',
	'dom'
);

$missing_extensions = array();
foreach( $required_extensions as $required_extension ) {
	if( ! extension_loaded($required_extension) ) {
		$missing_extensions[] = $required_extension;
	}
}


?><!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8">
	<meta name="viewport" content="width=device-width,initial-scale=1.0">
	<title>🏡 Knot Installer</title>

	<style>
		fieldset {
			margin-top: 1.5em;
		}
			fieldset legend {
				font-weight: bolder;
			}
	hr {
		margin: 2em 0;
		border: 0;
		border-top: 1px solid;
	}
	code {
		background: rgba(0,0,0,.1);
	}
	</style>
</head>
<body>
<main style="max-width: 600px; margin: 0 auto">

	<h1>🏡 Knot Installer</h1>

<?php

if( $local_phpversion[0] < $php_min_version_major ) {

	?>
	<p><strong>Error:</strong> your PHP version is too old (you need at least <code>PHP <?php echo $php_min_version_major; ?></code>).</p>
	<p>Please upgrade your PHP version to at least version <code><?php echo $php_min_version_major ?></code></p>
	<p>Then <a href="<?= $self ?>">refresh this page</a>.</p>
	<?php

} elseif( count($missing_extensions) ) {

	?>
	<p><strong>Error:</strong> it looks like your server is missing one or more extension that we depend on. Please make sure the following extensions are installed and working:</p>
	<ul>
		<?php
		foreach( $missing_extensions as $missing_extension ) {
			?>
			<li>
				<code><?php echo $missing_extension ?></code>
			</li>
			<?php
		}
		?>
	</ul>
	<p>Then <a href="<?= $self ?>">refresh this page</a>.</p>
	<?php

} elseif( $is_installed ) {

	?>
	<p>It looks like a Knot System is already installed at this location!</p>
	<p>Please delete the existing subfolders, then <a href="<?= $self ?>">refresh this page</a>.</p>
	<?php

} elseif( ! isset($_POST['action'])
 || $_POST['action'] != 'install'
 || empty($_POST['knot-site'])
 || empty($_POST['knot-site']['author_name'])
 || empty($_POST['knot-site']['site_title'])
) {

	?>

	<p>This script will install all modules required for a full <strong>Knot System</strong> installation. These modules are:</p>
	<ul>
		<li><strong>Knot Site</strong> as the website (and micropub server), that visitors will see</li>
		<li><strong>Knot Home</strong> as the micropub & microsub client, where you can write new posts and read posts from websites you follow</li>
		<li><strong>Knot Daemon</strong> as the microsub server, that will manage the websites you follow and collect new posts they publish</li>
		<li><strong>Knot Auth</strong> as the IndieAuth server, that will authorize you via a password, when you want to log in</li>
		<li><strong>Knot Control</strong> as a configuration interface where you can change some config options and search for updates</li>
	</ul>

	<p>Your server should meet all the necessary requirements.</p>

	<form id="install_form" method="POST" action="<?= $self ?>">

		<fieldset>
			<legend>Settings</legend>
	
			<p>You need to provide some basic information before we can begin the installation:</p>

			<p><label><strong>Version</strong><br><select name="version">
				<option value="latest" selected>latest stable release (recommended)</option>
				<option value="dev">unstable dev release (not recommended)</option>
			</select></label></p>

			<p><label><strong>Site Title</strong><br><input type="text" name="knot-site[site_title]" required><br><small>(the title of your website)</small></label></p>
			<p><label><strong>Author Name</strong><br><input type="text" name="knot-site[author_name]" required><br><small>(your name, displayed on your website)</small></label></p>
			<p><label id="password-label"><strong>Password</strong><br><input type="text" class="password-field" name="knot-auth[password]" minlength="8" required> <span class="password-toggle" style="display: none;"></span><br><small>(you use this password to log into the backend, where you can write new posts and read posts from pages you follow; the password needs to be at least 8 characters long)</small></label></p>
			<p><label><input type="checkbox" name="knot-site[testcontent]" value="true" checked> create Knot Site test content<br><small>(add some test content to your website, so you can check that everything works; this is optional)</small></label>

			<script>
			(function(){
				var label = document.getElementById('password-label'),
					field = label.querySelector('.password-field'),
					toggle = label.querySelector('.password-toggle');

				toggle.style.display = 'inline-block';
				field.type = 'password';
				toggle.innerHTML = '(<a href="#">show password</a>)';

				toggle.addEventListener( 'click', function(){
					if( field.type == 'password' ) {
						field.type = 'text';
						toggle.innerHTML = '(<a href="#">hide password</a>)';
					} else {
						field.type = 'password';
						toggle.innerHTML = '(<a href="#">show password</a>)';
					}
				});
			})();
			</script>

		</fieldset>

		<p><strong>This is an early beta version!</strong> Some things may break, or change in the future!</p>

		<p>Your website will be installed at <strong><?= $baseurl ?></strong>. Please make sure that this is correct.</p>

		<input type="hidden" name="action" value="install">
		<button>start installation</button>

	</form>

	<script>
	(function(){
		var form = document.getElementById('install_form');
		form.addEventListener('submit', function(e){
			var button = form.querySelector('button');
			button.disabled = true;
		});
	})();
	</script>

	<?php

} else {

	// running install:

	$temp_folder = $abspath.'tmp/';
	if( ! is_dir($temp_folder) ) {
		$oldumask = umask(0); // we need this for permissions of mkdir to be set correctly
		if( @mkdir( $temp_folder, 0774, true ) === false ) {
			$temp_folder = false;
		}
		umask($oldumask); // we need this after changing permissions with mkdir
	}

	if( ! $temp_folder ) {
		?>
		<p><strong>Error:</strong> could not create temporary <code>tmp/</code> folder. Please check the permissions of this directory.<br>The permission of the folder <em><?= $abspath ?></em> should be set to <code>644</code> (or <code>664</code>).</p>
		<p>After changing the permission, <a href="<?= $self ?>">refresh this page</a>.</p>
		<?php
		exit;
	}

	foreach( $sources as $source => $source_info ) {

		$version = 'stable';
		if( isset($_POST['version']) && $_POST['version'] == 'dev' ) {
			$version = 'dev';
		}

		flush();

		echo '<h3>Installing '.$source.'</h3>';

		$target_folder = $abspath.$source_info['target'];

		install_module( $source, $target_folder, $version );

		flush();

		time_nanosleep(0,500000000); // sleep for 0.5 seconds

		$config = [];
		if( ! empty($_POST[$source]) ) {
			$config = $_POST[$source];
		}


		// automatically set some additional options:
		$config['setup'] = true;
		if( $source == 'knot-site') {
			$config['baseurl_overwrite'] = $baseurl;
			$config['basefolder_overwrite'] = $basefolder;
			$config['microsub'] = $baseurl.$sources['knot-daemon']['target'];
			$config['indieauth-metadata'] = $baseurl.$sources['knot-auth']['target'].'metadata';
		} elseif( $source == 'knot-home' ) {
			$config['authorized_urls'] = $baseurl;
			$config['start'] = true;
		} elseif( $source == 'knot-daemon' ) {
			$config['authorized_urls'] = $baseurl;
			$config['refresh_on_connect'] = 'true';
		} elseif( $source == 'knot-auth' ) {
			$config['me'] = $baseurl;
		}


		// call setup of this module to create the config
		$setup_url = $baseurl.$source_info['target'].'index.php';

		echo '<p>updating config … ';
		flush();

		post_request( $setup_url, $config );

		time_nanosleep(0,500000000); // sleep for 0.5 seconds

		if( ! file_exists($abspath.$source_info['target'].'config.php') ) {
			echo '<span style="color: red;">could not create config file!</span>';
		} else {
			echo 'done.';
		}
		echo '</p>';

	}

	flush();

	$content = "# BEGIN knot\r\n<IfModule mod_rewrite.c>\r\nRewriteEngine on\r\nRewriteBase ".$basefolder."\r\n\r\nRewriteCond %{REQUEST_FILENAME} !-d\r\nRewriteCond %{REQUEST_FILENAME} !-f\r\nRewriteRule (.*) knot-site/$1 [L,QSA]\r\nRewriteRule ^$ knot-site/ [L,QSA]\r\n</IfModule>\r\n# END knot";
	file_put_contents( $abspath.'.htaccess', $content );

	echo '<h3>cleaning up</h3>';
	flush();

	delete_directory($temp_folder);

	unlink( $abspath.$self );

	echo '<p>all done.</p>';

	if( array_key_exists('knot-site', $sources) ) {
		echo '<p>&raquo; you can view your Knot Site at <a href="'.$baseurl.'" target="_blank" rel="noopener">'.$baseurl.'</a> - this is your personal website.</p>';
	}
	if( array_key_exists('knot-home', $sources) ) {
		echo '<p>&raquo; you can log into Knot Home at <a href="'.$baseurl.$sources['knot-home']['target'].'?login_url='.urlencode($baseurl).'" target="_blank" rel="noopener">'.$baseurl.$sources['knot-home']['target'].'</a> to write new posts and read posts from other websites.</p>';
	}
	if( array_key_exists('knot-control', $sources) ) {
		echo '<p>&raquo; you can log into Knot Control at <a href="'.$baseurl.$sources['knot-control']['target'].'?login_url='.urlencode($baseurl).'" target="_blank" rel="noopener">'.$baseurl.$sources['knot-control']['target'].'</a> to edit settings and update all modules.</p>';
	}

	flush();

}

?>
</main>
</body>
</html>
<?php


function post_request( $url, $post_data = array() ){
	global $useragent;

	$ch = curl_init( $url );
	curl_setopt( $ch, CURLOPT_RETURNTRANSFER, 1 );
	curl_setopt( $ch, CURLOPT_HEADER, true );
	curl_setopt( $ch, CURLOPT_USERAGENT, $useragent );
	curl_setopt( $ch, CURLOPT_POST, 1 );
	if( count($post_data) ) {
		curl_setopt( $ch, CURLOPT_POSTFIELDS, $post_data );
	}

	curl_setopt( $ch, CURLOPT_FOLLOWLOCATION, true );
	curl_setopt( $ch, CURLOPT_TIMEOUT, 10 );

	$response = curl_exec( $ch );

	curl_close( $ch );

	return $response;
}


function install_module( $source, $target, $version ) {

	global $temp_folder, $sources, $useragent;

	if( $version == 'dev' ) {

		$zipball = $sources[$source]['zipball_dev'];

		$zip_folder_name_start = $source.'-';

		echo '<p>Installing unstable dev release</p>';

	} else {

		$zipball = $sources[$source]['zipball_stable'];

		$json = get_remote_json( $zipball );
		if( ! $json || ! is_array($json) || ! count($json) ) {
			?>
			<p><strong>Error:</strong> could not get release information from GitHub</p>
			<?php
			return;
		}

		$latest_release = $json[0];

		if( empty($latest_release->zipball_url) ) {
			?>
			<p><strong>Error:</strong> could not get zip download url from GitHub</p>
			<?php
			exit;	
		}
		$zipball = $latest_release->zipball_url;

		$zip_folder_name_start = 'knot-system-'.$source.'-';

		echo '<p>Installing latest stable release</p>';

	}

	echo '<p>Downloading .zip from GitHub … ';
	flush();

	$temp_zip_file = $temp_folder.$source.'.zip';
	if( file_exists($temp_zip_file) ) unlink($temp_zip_file);

	$file_handle = fopen( $temp_zip_file, 'w+' );

	$ch = curl_init( $zipball );
	curl_setopt( $ch, CURLOPT_FOLLOWLOCATION, 1 );
	curl_setopt( $ch, CURLOPT_RETURNTRANSFER, 1 );
	curl_setopt( $ch, CURLOPT_USERAGENT, $useragent );
	curl_setopt( $ch, CURLOPT_FILE, $file_handle );
	curl_exec( $ch );
	curl_close( $ch );
	fclose($file_handle);

	echo 'done.</p>';

	echo '<p>Extracting .zip file … ';
	flush();


	$module_temp_folder = $temp_folder.$source.'/';
	if( is_dir($module_temp_folder) ) delete_directory($module_temp_folder);
	$oldumask = umask(0); // we need this for permissions of mkdir to be set correctly
	mkdir( $module_temp_folder );
	umask($oldumask); // we need this after changing permissions with mkdir

	$zip = new ZipArchive;
	$res = $zip->open($temp_zip_file);
	if( $res !== TRUE ) {
		echo '<p><strong>Error:</strong> could not extract .zip file</p>';
		return;
	}
	$zip->extractTo( $module_temp_folder );
	$zip->close();

	echo 'done.</p>';

	$subfolder = false;
	foreach( scandir( $module_temp_folder ) as $obj ) {
		if( $obj == '.' || $obj == '..' ) continue;
		if( ! is_dir($module_temp_folder.$obj) ) continue;
		if( ! str_starts_with($obj, $zip_folder_name_start) ) continue;
		// the zip file should have exactly one subfolder. this is what we want to get here
		$subfolder = $module_temp_folder.$obj.'/';
	}

	if( ! $subfolder ) {
		echo '<p><strong>Error:</strong> something went wrong with the .zip file</p>';
		return;
	}

	echo '<p>Moving files to new location … ';

	flush();

	move_folder_to( $subfolder, $target );

	echo 'done.</p>';

	flush();

}


function move_folder_to( $source, $target ){
    if( ! is_dir($target) ) {
    	$oldumask = umask(0); // we need this for permissions of mkdir to be set correctly
    	mkdir( $target, null, true );
    	umask($oldumask); // we need this after changing permissions with mkdir
    }
    rename( $source, $target );
}


function get_remote_json( $url ) {

	global $installer_version, $useragent;

	$ch = curl_init( $url );
	curl_setopt( $ch, CURLOPT_RETURNTRANSFER, 1 );
	curl_setopt( $ch, CURLOPT_USERAGENT, $useragent );
	$response = curl_exec($ch);
	curl_close( $ch );

	$json = json_decode($response);

	return $json;
}


function delete_directory( $dirPath ) {

	if( ! is_dir($dirPath) ) return;

	$objects = scandir($dirPath);
	foreach ($objects as $object) {
		if( $object == "." || $object == "..") continue;

		if( is_dir($dirPath . DIRECTORY_SEPARATOR . $object) ){
			delete_directory($dirPath . DIRECTORY_SEPARATOR . $object);
		} else {
			unlink($dirPath . DIRECTORY_SEPARATOR . $object);
		}
	}
	rmdir($dirPath);
}
